File: Haskell/SampleHaskell/cps.hs
add :: Int -> Int -> Int
add x y = x + y

square :: Int -> Int
square x = x * x

pythagoras :: Int -> Int -> Int
pythagoras x y = add (square x) (square y)

add_cps :: Int -> Int -> ((Int -> r) -> r)
add_cps x y = \k -> k (add x y)

square_cps :: Int -> ((Int -> r) -> r)
square_cps x = \k -> k (square x)

pythagoras_cps :: Int -> Int -> ((Int -> r) -> r)
pythagoras_cps x y = \k ->
    square_cps x $ \x_squared ->
    square_cps y $ \y_squared ->
    add_cps x_squared y_squared $ k
    
chainCPS :: ((a -> r) -> r) -> (a -> ((b -> r) -> r)) -> ((b -> r) -> r)
chainCPS s f = \k -> s $ \x -> f x $ k

pythagoras_cps' :: Int -> Int -> ((Int -> r) -> r)
pythagoras_cps' x y =
    square_cps x `chainCPS` \x_squared ->
    square_cps y `chainCPS` \y_squared ->
    add_cps x_squared y_squared

main = do
    print $ pythagoras 3 4
    pythagoras_cps 3 4 print
    pythagoras_cps' 3 4 print

{-
25
25
25
-}

File: Java/Samples/src/main/java/fp/CPS.java
package fp;

import java.util.function.IntConsumer;

public class CPS {
	static int add(int x, int y) {return x + y;}
	static int square(int x) {return x * x;}
	static int pythagoras(int x, int y){return add(square(x), square(y));}
	static void add_cps(int x, int y, IntConsumer f) {f.accept(add(x, y));}
	static void square_cps(int x, IntConsumer f) {f.accept(square(x));}
	static void pythagoras_cps(int x, int y, IntConsumer f) {
		square_cps(x, x_squared ->
		square_cps(y, y_squared ->
		add_cps(x_squared, y_squared, f)));
	}
	public static void main(String[] args) {
		System.out.println(pythagoras(3, 4));
		pythagoras_cps(3, 4, x -> System.out.println(x));
	}
}


